[DOC-11397] Document recursive queries using WITH clause

julia-cb opened the pull request:
Preview page: [Recursive WITH Clause](https://preview.docs-test.couchbase.com/recursivequery/server/current/n1ql/n1ql-language-reference/recursive.html)

Jira ticket: [DOC-11397](https://issues.couchbase.com/browse/DOC-11397)

Changes in this PR:

- New page called "Recursive WITH clause"
- New ebnf diagrams
- Google doc with dev information: https://docs.google.com/document/d/18ospWJSP3-P-SqI5LcMBR2z7tTQhEbmyLN3kfEQ4RtQ/edit

(Again, please ignore the Eventing page that was deleted! Not sure why it keeps being pushed with my query PRs)

julia-cb committed:
Updated Introduction to Eventing page

julia-cb committed:
Created new recursive file

julia-cb committed:
Generated png files

julia-cb committed:
Removed eventing page I pushed by accident - oops

julia-cb committed:
Updated recursive page and added new ebnf diagram

julia-cb committed:
Updated example 3

julia-cb committed:
Updated purpose section

julia-cb commented:
@simon-dew Hi Simon, the developer google doc mentions adding the recursive query info to the existing "WITH clause" page but there was so much content I created a new, separate page for it. Please let me know if that was the right call!

GauravJayaraj reviewed (commented):
Hi,
Looks good:) , can we add the examples with smaller datasets? maybe only the employee hierarchy one for the benefit of the user understanding usage ( but yeah no issues if aren't adding it)

Also can you please let me know how I can access https://preview.docs-test.couchbase.com/recursivequery/server/current/n1ql/n1ql-language-reference/recursive.html

RichardSmedley commented:
@GauravJayaraj 

> Also can you please let me know how I can access https://preview.docs-test.couchbase.com/recursivequery/server/current/n1ql/n1ql-language-reference/recursive.html

https://couchbasecloud.atlassian.net/wiki/spaces/DOCS/pages/1971224949/Docs+Team+demo+site+passwords

GauravJayaraj commented on modules/n1ql/assets/images/n1ql-language-reference/recursive-with-clause.png:
Rename ```rselect``` non-terminal to ```recursive-with-clause-select``` as that would be more descriptive

GauravJayaraj commented on modules/n1ql/pages/n1ql-language-reference/recursive.adoc:
```diff
@@ -0,0 +1,315 @@
+image::n1ql-language-reference/recursive-with-clause.png["Syntax diagram", align=left]
+
+----
+include::partial$grammar/dql.ebnf[tag=recursive-with-clause-union]
```
the new non-terminal -> ```recursive-with-clause-select```

Also we are **missing** the syntax diagrams for ```anchor-clause``` and ```recursive-clause```

anchor-clause would be a ```select```  from here: https://docs.couchbase.com/server/current/n1ql/n1ql-language-reference/select-syntax.html 

recursive-clause would be a ```select-term``` also from the above page.

julia-cb replied:
Hi Gaurav, do you mean we should add `recursive-with-clause-union ::= anchor_clause ( 'UNION' | 'UNION ALL' )? recursive_clause` to the "SELECT Syntax" page?

GauravJayaraj replied:
Hi,
No, the sorry for the confusion with the syntax diagram
attaching an updated version -> https://drive.google.com/file/d/14_ZlLF2EDZxjZr9dyMWr2P3aAqFi3jX_/view?usp=sharing 

here the anchor clause and recursive clause rules
anchor_clause::= select  // the select can be hyperlinked to the select over here https://docs.couchbase.com/server/current/n1ql/n1ql-language-reference/select-syntax.html 

similarly , for recursive_clause ::= select-term // select-term can be hyperlinked to select-term here https://docs.couchbase.com/server/current/n1ql/n1ql-language-reference/select-syntax.html 

Also mention, select-term in the recursive clause will be marked as recursive only if it refers to the with-alias...

simon-dew replied:
This comment is mostly superseded by my suggested changes to the EBNF. We need to strike a balance between legibility and brevity 😉 

Incidentally, we _will_ have to update the SELECT syntax page, but let's leave that outside the scope of this PR for now.

julia-cb replied:
Sounds good! Thank you both!

GauravJayaraj commented on modules/n1ql/pages/n1ql-language-reference/recursive.adoc:
```diff
@@ -0,0 +1,315 @@
+----
+====
+
+.Example finding all possible routes leaving from the TLV airport and going to 2 other airports
```
Add this before, as recursive queries tend to take long without appropriate indexes.
CREATE INDEX srcdestidx ON `travel-sample`.inventory.route( sourceairport, destinationairport );

simon-dew replied:
Check whether an index like this already exists in travel sample

GauravJayaraj replied:
Hi @simon-dew ,
I can confirm that an index -> def_inventory_route_route_src_dst_day :
Definition  ```CREATE INDEX `def_inventory_route_route_src_dst_day` ON `travel-sample`.`inventory`.`route`(`sourceairport`,`destinationairport`,(distinct (array (`v`.`day`) for `v` in `schedule` end))) WITH { "defer_build":true }```

This index is used by the query but the runtime is roughly 14seconds on my machine

But the above index-> ```CREATE INDEX srcdestidx ON travel-sample.inventory.route( sourceairport, destinationairport );```

Results in a faster query( 5seconds something). I would prefer that this index be created:)

GauravJayaraj commented on modules/n1ql/pages/n1ql-language-reference/recursive.adoc:
```diff
@@ -0,0 +1,315 @@
+        recroute.destinationairport = r.sourceairport
+    WHERE r.sourceairport!="TLV" and recroute.depth<2 
+)
+SELECT * FROM recroute LIMIT 10;
```
LIMIT BY 10 defeats the purpose of the example as, all results would indicate traversal of depth 0 or only the root of the graph

After trying again I can confirm that it takes atleast 2000docs to see results from an iteration of depth = 2.

Which is why I would show simpler examples shown in the dev google doc.
But add the travel-sample example at the end just for reference.
Also those example indicate usage of the OPTIONS and CYCLE subclause.

you can see here how snowflake have done a similar job with a smaller example allowing user to first setup the data using insert and later a query to show usage.
https://docs.snowflake.com/en/user-guide/queries-cte#recursive-ctes-and-hierarchical-data

simon-dew replied:
I agree with this comment. Note that if you don't want to show the entire result set, to save space in the document, you can cut down the example using an ellipsis `...`, like you did in the LATERAL examples.

julia-cb committed:
Rebase branch

simon-dew commented on modules/n1ql/partials/grammar/dql.ebnf:
```diff
@@ -45,6 +45,26 @@ alias ::= identifier
+recursive-with-clause-cycle ::= 'CYCLE' fieldnames_list* 'RESTRICT'
+/* end::recursive-with-clause-cycle[] */
+
+
```
I'm all in favour of descriptive names in the syntax diagrams but that has to be balanced with brevity 😉 

Corrected syntax following the [N1QL parser](https://github.com/couchbase/query/blob/trinity/parser/n1ql/n1ql.y) 

```suggestion
/* WITH RECURSIVE Clause */

/* tag::with-recursive-clause[] */
with-recursive-clause ::= 'WITH' 'RECURSIVE' alias 'AS' '(' ( recursive-select | select | expr ) ')' cycle-clause? options-clause?
                                       ( ',' alias 'AS' '(' ( recursive-select | select | expr ) ')' cycle-clause? options-clause? )*
/* end::with-recursive-clause[] */

/* tag::recursive-select[] */
recursive-select ::= anchor-select ('UNION' | 'UNION ALL') recursive-select-term
/* end::recursive-select[] */

/* tag::cycle-clause[] */
cycle-clause ::= 'CYCLE' expr ( ',' expr )* 'RESTRICT'
/* end::cycle-clause[] */

/* tag::options-clause[] */
options-clause ::= 'OPTIONS' expr
/* end::options-clause[] */

/* tag::anchor-select[] */
anchor-select ::= select
/* end::anchor-select[] */

/* tag::recursive-select-term[] */
recursive-select-term ::= select-term
/* end::recursive-select-term[] */
```

simon-dew commented on modules/n1ql/assets/images/n1ql-language-reference/recursive-with-clause-cycle.png:
Delete and replace with new diagrams from `dql.ebnf`

simon-dew commented on modules/n1ql/assets/images/n1ql-language-reference/recursive-with-clause-options.png:
Delete and replace with new diagrams from `dql.ebnf`

simon-dew commented on modules/n1ql/assets/images/n1ql-language-reference/recursive-with-clause-union.png:
Delete and replace with new diagrams from `dql.ebnf`

simon-dew commented on modules/n1ql/assets/images/n1ql-language-reference/recursive-with-clause.png:
Delete and replace with new diagrams from `dql.ebnf`

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/recursive.adoc:
```diff
@@ -0,0 +1,315 @@
+:description: Use the RECURSIVE keyword to enable recursive referencing in WITH clause common table expressions.
+:imagesdir: ../../assets/images
+:page-topic-type: reference
+
```
Boilerplate text for new topic in this release:

```suggestion
:page-status: Couchbase Server 7.6

```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/recursive.adoc:
```diff
@@ -0,0 +1,315 @@
+= Recursive WITH Clause
+:description: Use the RECURSIVE keyword to enable recursive referencing in WITH clause common table expressions.
```
I would probably call it the WITH RECURSIVE clause, following the order of keywords in the syntax?

```suggestion
= WITH RECURSIVE Clause
:description: Use the WITH RECURSIVE clause to enable recursive referencing in common table expressions.
```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/recursive.adoc:
```diff
@@ -0,0 +1,315 @@
+
+The common table expressions (CTEs) created by the WITH clause simplify complex queries and create temporary result sets which can be used as data sources or as expressions for queries.
+
+The SELECT definition for CTEs is split by the UNION/UNION ALL operators.
+
+* The left arm of UNION/UNION ALL contains an anchor clause, which is non-recursive. 
+The anchor clause produces an initial set of documents as an intermediate result set for the CTE keyspace.
+
+* The right arm of UNION/UNION ALL contains the recursive clause.
+The recursive clause produces a new set of documents.
```
Reworking this to introduce the concepts and why we would need to use this.

```suggestion
The WITH RECURSIVE clause is an extension of the WITH clause which enables you to create recursive queries.
You can use recursive queries with hierarchical data, or data in a tree structure.
In these cases, the data may have an arbitrary number of levels, and you won't know in advance how deeply you need to traverse the data.

The CTE for a recursive query includes a UNION or UNION ALL xref:n1ql:n1ql-language-reference/union.adoc[set operator]. 

* The left arm of UNION/UNION ALL contains an anchor clause, which is non-recursive. 
The anchor clause produces an initial set of documents as an intermediate result set for the CTE keyspace.

* The right arm of UNION/UNION ALL contains the recursive clause.
The recursive clause produces a new set of documents.
```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/recursive.adoc:
```diff
@@ -0,0 +1,315 @@
+[#syntax]
+== Syntax
+
+[source,ebnf]
+----
+include::partial$grammar/dql.ebnf[tag=recursive-with-clause]
+----
+
+image::n1ql-language-reference/recursive-with-clause.png["Syntax diagram", align=left]
```
```suggestion
[source,ebnf]
----
include::partial$grammar/dql.ebnf[tag=with-recursive-clause]
----

image::n1ql-language-reference/with-recursive-clause.png["Syntax diagram", align=left]
```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/recursive.adoc:
```diff
@@ -0,0 +1,315 @@
+
+image::n1ql-language-reference/recursive-with-clause.png["Syntax diagram", align=left]
+
+----
+include::partial$grammar/dql.ebnf[tag=recursive-with-clause-union]
+----
+
+image::n1ql-language-reference/recursive-with-clause-union.png["Syntax diagram UNION/UNION ALL", align=left]
```
```suggestion
[[recursive-select]]
=== Recursive SELECT
----
include::partial$grammar/dql.ebnf[tag=recursive-select]
----

image::n1ql-language-reference/recursive-select.png["Syntax diagram", align=left]

The definition for a recursive CTE must be a SELECT statement that includes a UNION or UNION ALL set operator.
If it isn't, the CTE is treated as a non-recursive CTE.

anchor-select::
A xref:n1ql:n1ql-language-reference/select-syntax.adoc[SELECT query] that represents the anchor clause of the CTE.

recursive-select-term::
A xref:n1ql:n1ql-language-reference/select-syntax.adoc#select-term[select term] that represents the recursive clause of the CTE.
```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/recursive.adoc:
```diff
@@ -0,0 +1,315 @@
+
+image::n1ql-language-reference/recursive-with-clause-union.png["Syntax diagram UNION/UNION ALL", align=left]
+
+[#arguments]
+== Arguments
+
+The following arguments are optional.
+
+[source,ebnf]
+----
+include::partial$grammar/dql.ebnf[tag=recursive-with-clause-options]
+----
+
+image::n1ql-language-reference/recursive-with-clause-options.png["Argument diagram", align=left]
+
+options:: Object that signals an early exit.
+* `levels`: An exit after level `N`. 
+This object assumes the Anchor is at level 0.
+* `documents`: An exit after accumulating `N` documents. 
+
+[source,ebnf]
+----
+include::partial$grammar/dql.ebnf[tag=recursive-with-clause-cycle]
+----
+
+image::n1ql-language-reference/recursive-with-clause-cycle.png["Argument diagram", align=left]
+
+cycle:: `CYCLE <fieldnames_list> RESTRICT`. 
+List of field names in which you want to perform cycle detection.
```
Use individual 3rd level headings for each subclause in the syntax, rather than a catch-all "Arguments" section. Do the CYCLE clause first, because it comes first in the syntax. Use a table to describe the schema for the OPTIONS object.

Note also that we don't need to include any more than minimal alt text for the syntax diagrams; the EBNF text _is_ the alt text for the syntax diagram.

```suggestion
[[cycle-clause]]
=== CYCLE Clause

[source,ebnf]
----
include::partial$grammar/dql.ebnf[tag=cycle-clause]
----

image::n1ql-language-reference/cycle-clause.png["Syntax diagram", align=left]

The optional CYCLE clause provides one method for avoiding infinite recursions.
It enables you to specify one or more fields whose values are likely to repeat.

expr::
(Required) An identifier representing a field.

[[options-clause]]
=== OPTIONS Clause
[source,ebnf]
----
include::partial$grammar/dql.ebnf[tag=options-clause]
----

image::n1ql-language-reference/options-clause.png["Syntax diagram", align=left]

The optional OPTIONS clause provides another method for avoiding infinite recursions.
It enables you to specify that the recursion should exit after a specified level, or after accumulating a specified number of documents.

expr::
(Required) An object with the following properties.

[options="header", cols="~a,~a,~a"]
|===
|Name|Description|Schema

|**levels** +
__optional__
|Recursion should exit after reaching this level.
This assumes the anchor is at level 0.
|Integer

|**documents** +
__optional__
|Recursion should exit after accumulating this many documents.
|Integer
|===
```

simon-dew reviewed (changes requested):
Hi @julia-cb &mdash; a great start. Sorry, there are some fairly big comments here. This is probably enough to get started with.

julia-cb committed:
Apply suggestions from code review

Co-authored-by: Simon Dew <39966290+simon-dew@users.noreply.github.com>

simon-dew reviewed (approved):
Thanks @julia-cb ! If you've finished making edits, I would say let's get this merged 👍 

I will have some edits to make to the SELECT syntax pages in a separate PR, but I'll keep you in the loop on that so you can see what I've done

GauravJayaraj commented:
@julia-cb Looks good 😇, have we decided not to add the smaller examples??
+ missing syntax diagrams for anchor clause, recursive clause as well as grammar.

Also can you update the preview page again, I'd like to see the preview once😅
Thanks.

GauravJayaraj commented on modules/n1ql/assets/images/n1ql-language-reference/recursive-with-clause-options.png:
Need to update this to 
OPTIONS object

simon-dew replied:
This has been addressed by the most recent changes to the EBNF

julia-cb committed:
Updated code example and renamed recursive.adoc to with-recursive.adoc

julia-cb committed:
Update recursive.adoc to with-recursive.adoc

julia-cb committed:
Recreated diagrams

julia-cb committed:
Deleted old diagrams

julia-cb committed:
Added OPTIONS and CYCLE examples

julia-cb commented on modules/n1ql/pages/n1ql-language-reference/with-recursive.adoc:
```diff
@@ -0,0 +1,358 @@
+.Examples using the OPTIONS clause and the CYCLE clause to avoid infinite recursion
+====
+.Example resulting in infinite recursion
+[source,sqlpp]
+----
+WITH RECURSIVE similar_items AS (
+  SELECT p.*, 0 as lvl FROM products p WHERE p.id=1
+    UNION
+  SELECT p1.*, similar_items.lvl+1 as lvl FROM products p1,
+    similar_items WHERE p1.id IN similar_items.related_items
+)
+SELECT * FROM similar_items;
+----
+
+.Solution using the OPTIONS clause to add a level limit
+[source,sqlpp]
+----
+WITH RECURSIVE similar_items AS (
+  SELECT p.*, 0 as lvl FROM products p WHERE p.id=1
+    UNION
+  SELECT p1.*, similar_items.lvl+1 as lvl FROM products p1,
+    similar_items WHERE p1.id IN similar_items.related_items
+)
+OPTIONS {"levels:2"}
+SELECT * FROM similar_items;
+---- 
+
+.Solution using the CYCLE clause to track which fields are likely to repeat and cause a cycle
+[source,sqlpp]
+----
+WITH RECURSIVE similar_items AS (
+  SELECT p.*, 0 as lvl FROM products p WHERE p.id=1
+    UNION
+  SELECT p1.*, similar_items.lvl+1 as lvl FROM products p1,
+    similar_items WHERE p1.id IN similar_items.related_items
+)
+CYCLE id RESTRICT
+SELECT * FROM similar_items;
+---- 
+====
```
@simon-dew Hey Simon, I added these examples here to show how to use OPTIONS and CYCLE to avoid causing infinite recursion. Let me know what you think! :)

julia-cb commented:
@GauravJayaraj Hi Gaurav, here's the updated preview page: https://preview.docs-test.couchbase.com/with-recursive-example/server/current/n1ql/n1ql-language-reference/with-recursive.html

Here are the changes I made:

- Updated the text and diagrams with Simon's suggestions
- Updated the travel-sample response to include the different depths
- Added the OPTIONS and CYCLE examples from your google doc

Let me know what you think! :)

GauravJayaraj commented:
> @GauravJayaraj Hi Gaurav, here's the updated preview page: https://preview.docs-test.couchbase.com/with-recursive-example/server/current/n1ql/n1ql-language-reference/with-recursive.html
> 
> Here are the changes I made:
> 
> * Updated the text and diagrams with Simon's suggestions
> * Updated the travel-sample response to include the different depths
> * Added the OPTIONS and CYCLE examples from your google doc
> 
> Let me know what you think! :)

Hi, page looks good, just the last one thing(comment about the data for the examples). But otherwise good to merge:)

julia-cb commented:
Merging and closing this PR - I can make any more necessary edits to the page later. :)
